istioctl/cmd/injector-list.go

```diff
@@ -0,0 +1,311 @@
+		if len(namespaces) == 0 {
+			fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", hook.Name, revision, "DOES NOT AUTOINJECT", injectedImages[revision])
+			continue
+		}
+		for _, namespace := range namespaces {
+			fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", hook.Name, revision, namespace.Name, injectedImages[revision])
+		}
+	}
+	return w.Flush()
+}
+
+func getInjector(namespace *v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration) *admit_v1.MutatingWebhookConfiguration {
+	// find matching hook
+	for _, hook := range hooks {
+		for _, webhook := range hook.Webhooks {
+			nsSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector)
+			if err != nil {
+				continue
+			}
+			// @@@ nsSelector := api_pkg_labels.SelectorFromSet(webhook.NamespaceSelector)
+			if nsSelector.Matches(api_pkg_labels.Set(namespace.ObjectMeta.Labels)) {
+				// @@@ TODO verify the meaning of Webhooks is "OR" not "AND"
+				return &hook
+			}
+		}
+	}
+	return nil
+}
+
+func getInjectedRevision(namespace *v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration) string {
+	injector := getInjector(namespace, hooks)
+	if injector != nil {
+		return injector.ObjectMeta.GetLabels()[label.IstioRev]
+	}
+	newRev := namespace.ObjectMeta.GetLabels()[label.IstioRev]
+	oldLabel, ok := namespace.ObjectMeta.GetLabels()[analyzer_util.InjectionLabelName]
+	// If there is no istio-injection=disabled and no istio.io/rev, the namespace isn't injected
+	if newRev == "" && (ok && oldLabel == "disabled" || !ok) {
+		return ""
+	}
+	if newRev != "" {
+		return fmt.Sprintf("MISSING/%s", newRev)
+	}
+	return fmt.Sprintf("MISSING/%s", analyzer_util.InjectionLabelName)
+}
+
+func getMatchingNamespaces(hook *admit_v1.MutatingWebhookConfiguration, namespaces []v1.Namespace) []v1.Namespace {
+	retval := make([]v1.Namespace, 0)
+	for _, webhook := range hook.Webhooks {
+		nsSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector)
+		if err != nil {
+			return retval
+		}
+
+		for _, namespace := range namespaces {
+			if nsSelector.Matches(api_pkg_labels.Set(namespace.Labels)) {
+				// @@@ TODO Verify "OR" not "AND" on Webhooks
+				retval = append(retval, namespace)
+			}
+		}
+	}
+	return retval
+}
+
+func getPods(client kube.ExtendedClient) (map[resource.Namespace][]v1.Pod, error) {
+	retval := map[resource.Namespace][]v1.Pod{}
+	// All pods in all namespaces
+	pods, err := client.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return retval, err
+	}
+	for _, pod := range pods.Items {
+		podList, ok := retval[resource.Namespace(pod.GetNamespace())]
+		if !ok {
+			retval[resource.Namespace(pod.GetNamespace())] = []v1.Pod{pod}
+		} else {
+			retval[resource.Namespace(pod.GetNamespace())] = append(podList, pod)
+		}
+	}
+	return retval, nil
+}
+
+// getInjectionImages() returns a map of revision->dockerimage
+func getInjectionImages(client kube.ExtendedClient) (map[string]string, error) {
+	retval := map[string]string{}
```

Comment:
nit: `retval` > `images| revision` as it will return a map of `revision->dockerimage`

Reply:
I don't understand.

Reply:
Sorry I mean There should be a better name than `retval`. maybe `images` or `imageRevision`?